src/main/java/io/github/cottonmc/cotton/gui/impl/ScreenNetworkingImpl.java

```diff
@@ -58,51 +89,48 @@ public void send(Identifier message, Consumer<PacketByteBuf> writer) {
 		Objects.requireNonNull(writer, "writer");
 
 		PacketByteBuf buf = PacketByteBufs.create();
-		buf.writeVarInt(description.syncId);
-		buf.writeIdentifier(message);
 		writer.accept(buf);
-		description.getPacketSender().sendPacket(side == NetworkSide.SERVER ? SCREEN_MESSAGE_S2C : SCREEN_MESSAGE_C2S, buf);
+		ScreenMessageData data = new ScreenMessageData(description.syncId, message, buf);
+		description.getPacketSender().sendPacket(side == NetworkSide.SERVER ? new S2CScreenMessage(data) : new C2SScreenMessage(data));
 	}
 
 	public static void init() {
-		ServerPlayNetworking.registerGlobalReceiver(SCREEN_MESSAGE_C2S, (server, player, networkHandler, buf, responseSender) -> {
-			handle(server, player, buf);
+		PayloadTypeRegistry.playS2C().register(S2CScreenMessage.PACKET_ID, S2CScreenMessage.PACKET_CODEC);
+		PayloadTypeRegistry.playC2S().register(C2SScreenMessage.PACKET_ID, C2SScreenMessage.PACKET_CODEC);
+		ServerPlayNetworking.registerGlobalReceiver(C2SScreenMessage.PACKET_ID, (payload, context) -> {
+			handle(context.player().server, context.player(), payload.data());
 		});
 	}
 
-	public static void handle(Executor executor, PlayerEntity player, PacketByteBuf buf) {
+	public static void handle(Executor executor, PlayerEntity player, ScreenMessageData data) {
 		ScreenHandler screenHandler = player.currentScreenHandler;
 
-		// Packet data
-		int syncId = buf.readVarInt();
-		Identifier messageId = buf.readIdentifier();
-
 		if (!(screenHandler instanceof SyncedGuiDescription)) {
 			LOGGER.error("Received message packet for screen handler {} which is not a SyncedGuiDescription", screenHandler);
 			return;
-		} else if (syncId != screenHandler.syncId) {
-			LOGGER.error("Received message for sync ID {}, current sync ID: {}", syncId, screenHandler.syncId);
+		} else if (data.syncId() != screenHandler.syncId) {
+			LOGGER.error("Received message for sync ID {}, current sync ID: {}", data.syncId(), screenHandler.syncId);
 			return;
 		}
 
 		ScreenNetworkingImpl networking = instanceCache.get(screenHandler);
 
 		if (networking != null) {
-			MessageReceiver receiver = networking.messages.get(messageId);
+			MessageReceiver receiver = networking.messages.get(data.message());
 
 			if (receiver != null) {
-				buf.retain();
+				data.buf().retain();
```

Comment:
Since the buffer is now created by us, I think this extra `retain` might leak memory:
- Initial state: refcount = 1
- `retain`: refcount = 2
- `release`: refcount = 1 (needs 0 to be deallocated according to Netty `ReferenceCounted` docs)